### PR TITLE
[WIP] Penalty on shap calculation for higher variance

### PR DESCRIPTION
**Pull request by markdregan**
PR addresses #216 

**Overall objective:** Add penalty to features that has high variance in underlying shap values - when computing feature importance. This will (in theory) encourage selection of features that have more coherency across CV folds.

**Work tasks:**
- [ ] Add `shap_variance_penalty` parameter to `ShapRFECV.fit`, `compute` & `fit_compute`
- [ ] Add `shap_variance_penalty` parameter to `EarlyStoppingShapRFECV.fit`, `compute` & `fit_compute`
- [ ] Add `shap_variance_penalty` to `shap_helpers.calculate_shap_importance`
- [ ] Add `shap_variance_penalty` to `ShapModelInterpreter.fit`, `compute` & `fit_compute`
- [ ] Add to doc strings throughout
- [ ] Run simulation demonstrating impact of `shap_variance_penalty` feature

**Reviewers:** LMK any changes / improvements that can be made.

### CODE REVIEW

**Comment by ReinierKoops** [comment]
Thanks for the PR. I’ll review it some parts now, and again once it’s complete.

File: probatus/feature_elimination/feature_elimination.py [hunk 1]
```diff
@@ -526,12 +523,9 @@ def fit(
             # Get current dataset info
             current_features_set = remaining_features
             if columns_to_keep is None:
-                # Keeps the original order, while removing duplicate elements
-                remaining_removeable_features = pd.Series(current_features_set).unique()
+                remaining_removeable_features = list(set(current_features_set))
```
Comment (ReinierKoops) [inline probatus/feature_elimination/feature_elimination.py:526]: Set is by definition unordered. Using such implementation for classifiers (like Lightgbm) where the order of column input is important to determine which are kept and which are not, leads to bugs. Therefore we use pd.series.unique() which guarantees the order is the same as the input minus the features which are duplicate.

File: probatus/feature_elimination/feature_elimination.py [hunk 2]
```diff
@@ -526,12 +523,9 @@ def fit(
...
             else:
-                # Keeps the original order, while removing duplicate elements
-                remaining_removeable_features = pd.Series(list(current_features_set) + columns_to_keep).unique()
-
+                remaining_removeable_features = list(set(current_features_set) | set(columns_to_keep))
```
Comment (ReinierKoops) [inline probatus/feature_elimination/feature_elimination.py:528]: Same here

File: probatus/feature_elimination/feature_elimination.py [hunk 3]
```diff
@@ -556,7 +550,7 @@ def fit(
                     sample_weight=sample_weight,
                     **shap_kwargs,
                 )
-                for train_index, val_index in self.cv.split(current_X, self.y, groups)
+                for train_index, val_index in self.cv.split(current_X, self.y)
```
Comment (ReinierKoops) [inline probatus/feature_elimination/feature_elimination.py:553]: We need groups for the implementation since this guarantees a cv split which keeps the groups in mind (and thus conceptually allows us to alleviate some bias introduced when we don’t consider for these groups).

File: probatus/feature_elimination/feature_elimination.py [hunk 4]
```diff
@@ -565,19 +559,13 @@ def fit(
...
                 shap_importance_df, columns_to_keep=columns_to_keep
             )
-            # Ensures the order of the first list is kept as it was originally,
-            # while removing elements which are present in both lists.
-            remaining_features = np.setdiff1d(
-                pd.Series(current_features_set).unique(),
-                pd.Series(features_to_remove).unique(),
-                assume_unique=True,
-            )
+            remaining_features = list(set(current_features_set) - set(features_to_remove))
```
Comment (ReinierKoops) [inline probatus/feature_elimination/feature_elimination.py:568]: Again here

File: probatus/feature_elimination/feature_elimination.py [hunk 5]
```diff
@@ -918,7 +910,7 @@ def __init__(
                  and [LightGBM](https://lightgbm.readthedocs.io/en/latest/Parameters.html#metric-parameters).
                 Note that `eval_metric` is an argument of the model's fit method and it is different from `scoring`.
         """  # noqa
-        super().__init__(
+        super(EarlyStoppingShapRFECV, self).__init__(
```
Comment (ReinierKoops) [inline probatus/feature_elimination/feature_elimination.py:913]: This is not required.

File: probatus/feature_elimination/feature_elimination.py [hunk 6]
```diff
@@ -1249,32 +1245,6 @@ def _get_feature_shap_values_per_fold(self, X, y, clf, train_index, val_index, s
             val_index=val_index,
         )
 
-        # Due to deprecation issues (compatibility with Sklearn) set some params
```
Comment (ReinierKoops) [inline probatus/feature_elimination/feature_elimination.py:1252]: Out of curiosity, why is this removed?

**Review by ReinierKoops — CHANGES_REQUESTED** [review]

File: probatus/feature_elimination/feature_elimination.py (hunk 1 above)
Comment (markdregan) [reply probatus/feature_elimination/feature_elimination.py:526]: This is odd. The diff between my PR and main shown here is not right. I didn't make many of these changes. I think I know the issue. I think I based off an older version. I might have to do a fresh PR...

**Comment by markdregan** [comment]
@ReinierKoops - I seem to have based this PR off some older version of probatus. Just submitted fresh PR. Apologies. I'll delete this PR.
